teps/0089-spire-support.md

```diff
@@ -0,0 +1,184 @@
+This property is a critical feature, as it ensures that a TaskRun was created by Tekton.
+
+The basic flow would look like this:
+1. Tekton Pipelines receives a TaskRun config, and generates the Pod for it as usual
+1. Tekton Pipelines runs the TaskRun, and then requests a signature over the completed TaskRun yaml from SPIRE
+1. SPIRE has been set up to only issue signatures and SVIDs to the Tekton controller, which it does
+1. Tekton Pipelines stores the SVID and signature received from SPIRE as annotations on the TaskRun
+1. Tekton Chains verifies that SVID and signature against the TaskRun yaml
+1. If verification is successful, Chains proceeds normally. Otherwise, it stops and doesn't sign anything!
+
+With the SPIRE integration, we successfully mitigate the two issues mentioned above.
+Tekton Chains can be confident the TaskRun was created by Tekton, since only Tekton workloads can have SPIRE issued SVIDs.
+Tekton Chains can also verify that the TaskRun yaml wasn't modified by anybody other than the Tekton controller, because the signature exists over the entire TaskRun yaml.
+
+_NOTE: A cluster administrator could still potentially threaten this model by disabling SPIRE or changing settings on the k8s Workload Attestor. At SLSA 3, it is still allowed that some trusted administrators will have root access to the build system. This will eventually need to be addressed for SLSA 4_.
+
+
+## Motivation
+
+Security! =)
+
+We'll also need this feature for Tekton to achieve [SLSA 3](https://slsa.dev/levels), which requires Non-falsifiable Provenance.
+
+### Requirements
+* SPIRE is configured to only issue SVIDs to the Tekton controller via a [k8s Workload Attestor](https://github.com/spiffe/spire/blob/main/doc/plugin_agent_workloadattestor_k8s.md)
+
+### Goals
+
+* Once a TaskRun completes, Pipelines can request a signature over the TaskRun yaml from SPIRE
+* Pipelines can store the signature and SVID on the TaskRun itself as annotaitons
+* Chains should be able to verify a TaskRun wasn't modified after execution via the signature and SVID provided
+
+### Non-Goals
+* Protect against a malicious cluster admin (this will be a goal for SLSA 4)
+
+## SPIRE Concepts
+The Tekton Controller will send the TaskRun yaml (this is the `payload`) to SPIRE for signing.
+SPIRE will make sure that the request came from the Tekton controller, and once it has confirmed that it will sign the `payload`.
+SPIRE will return the following to the Tekton Controller:
+1. The `signature` over the `payload`
+1. A SPIFFE Verifiable Identity Document, or [SVID](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/#spiffe-verifiable-identity-document-svid), an X509 certificate containing the public key used to verify the `signature` against the `payload`
+
+## Proposal
+As mentioned above, the basic design looks like this:
+1. Tekton Pipelines receives a TaskRun config, and generates the Pod for it
```

Comment:
as an extra detail, related to my question above about modifications during execution, afaik the pipelines controller will be updating the TaskRun status as the TaskRun executes - I'm wondering if we want to be signing the TaskRun with each modification (and maybe even verifying that nothing has changed since the last modification??)

im not sure if im going too far but it seems like if we dont do that, something else could modify the taskrun status during execution, and the pipelines controller would just incrementally add to those changes, only signing at the end?

Reply:
that makes sense, and would probably be good to have. i'll add in some thoughts into the proposal!

Reply:
instead of signing the TaskRun with each modification, do you think it could make sense to store the current TaskRun in memory and compare it to the actual thing? maybe something like this:

1. store the initial TaskRun in memory
2. when we want to modify the running TaskRun, compare the running TaskRun to the one in memory.
3. if it doesn't match, then we never request an SVID and skip any SPIRE signing altogether
4. if it does match, then we modify the TaskRun and store the new modified version in memory, and repeat steps 2-4 until taskrun completes